File: UnityMcpBridge/Editor/Tools/AddClickStep.cs
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using UnityMcpBridge.Editor.Helpers;
using System.Text.RegularExpressions;
using System;

namespace UnityMcpBridge.Editor.Tools
{
    public static class AddClickStep
    {
        // Regular expressions for natural language parsing
        private static readonly Regex GraphPattern = new Regex(@"(?:in|graph|asset:?\s*)?([\w-]+\.asset)", RegexOptions.IgnoreCase);
        private static readonly Regex TargetPattern = new Regex(@"(?:in|on|at|button:?\s*)?([\w-]+)/([\w-]+)", RegexOptions.IgnoreCase);

        public static object HandleCommand(JObject @params)
        {
            try
            {
                // Add debug logging to see what parameters we're receiving
                Debug.Log($"AddClickStep received parameters: {@params?.ToString()}");

                string prompt = @params["prompt"]?.ToString();
                string targetGraphPath = @params["targetGraphPath"]?.ToString();
                string targetName = @params["targetName"]?.ToString();
                string stepId = @params["stepId"]?.ToString();

                Debug.Log($"Parsed parameters - targetGraphPath: {targetGraphPath}, targetName: {targetName}, prompt: {prompt}");

                // Try to parse from natural language if prompt is provided
                if (!string.IsNullOrEmpty(prompt))
                {
                    var graphMatch = GraphPattern.Match(prompt);
                    var targetMatch = TargetPattern.Match(prompt);

                    if (graphMatch.Success && string.IsNullOrEmpty(targetGraphPath))
                    {
                        targetGraphPath = $"Assets/_MCP/Graphs/{graphMatch.Groups[1].Value}";
                        Debug.Log($"Parsed targetGraphPath from prompt: {targetGraphPath}");
                    }

                    if (targetMatch.Success && string.IsNullOrEmpty(targetName))
                    {
                        targetName = $"{targetMatch.Groups[1].Value}/{targetMatch.Groups[2].Value}";
                        Debug.Log($"Parsed targetName from prompt: {targetName}");
                    }
                }

                // Validate required parameters
                if (string.IsNullOrEmpty(targetGraphPath))
                {
                    return Response.Error("targetGraphPath parameter is required.");
                }
                if (string.IsNullOrEmpty(targetName))
                {
                    return Response.Error("targetName parameter is required.");
                }

                // Check if the file exists before trying to load it
                if (!System.IO.File.Exists(targetGraphPath))
                {
                    return Response.Error($"File does not exist at path: {targetGraphPath}");
                }

                // Load the target graph asset
                var targetGraph = AssetDatabase.LoadAssetAtPath<StepsGraph>(targetGraphPath);
                if (targetGraph == null)
                {
                    return Response.Error($"Could not load StepsGraph from path: {targetGraphPath}. Make sure it's a valid StepsGraph asset.");
                }

                // Validate targetName format
                string[] targetParts = targetName.Split('/');
                if (targetParts.Length != 2)
                {
                    return Response.Error($"targetName must be in format 'prefabName/childName', got: {targetName}");
                }

                // Create the click step
                var clickStep = ScriptableObject.CreateInstance<ClickStep>();
                if (clickStep == null)
                {
                    return Response.Error("Failed to create ClickStep instance. Make sure ClickStep class exists and is properly defined.");
                }

                clickStep.stepId = stepId ?? System.Guid.NewGuid().ToString();
                
                // Set up the registry item for the target
                var registryItem = new RegistryItem
                {
                    prefabName = targetParts[0],
                    childName = targetParts[1]
                };
                clickStep._targetName = registryItem;

                Debug.Log($"Created ClickStep with ID: {clickStep.stepId}, targeting: {targetParts[0]}/{targetParts[1]}");

                // Initialize nodes list if it's null
                if (targetGraph.nodes == null)
                {
                    Debug.Log("Initializing nodes list for targetGraph");
                    targetGraph.nodes = new System.Collections.Generic.List<StepsGraphNode>();
                }

                // Add the step to the graph
                targetGraph.nodes.Add(clickStep);
                EditorUtility.SetDirty(targetGraph);
                
                // Force save and refresh
                AssetDatabase.SaveAssets();
                AssetDatabase.Refresh();

                Debug.Log($"Successfully added ClickStep to graph. Total nodes: {targetGraph.nodes.Count}");

                return Response.Success(new
                {
                    stepId = clickStep.stepId,
                    targetName = targetName,
                    graphPath = targetGraphPath,
                    nodesCount = targetGraph.nodes.Count,
                    message = "Click step added successfully"
                });
            }
            catch (Exception ex)
            {
                Debug.LogError($"Error in AddClickStep.HandleCommand: {ex.Message}\n{ex.StackTrace}");
                return Response.Error($"Exception occurred: {ex.Message}");
            }
        }
    }
}

File: UnityMcpServer/src/tools/add_click_step.py
from typing import Any, Dict
from mcp.server.fastmcp import Context, FastMCP
from unity_connection import get_unity_connection
import re

def register_add_click_step_tools(mcp: FastMCP):
    """Register add click step tools with the MCP server."""
    
    @mcp.tool()
    def add_click_step(
        ctx: Context,
        prompt: str = None,
        target_graph_path: str = None,
        target_name: str = None,
        step_id: str = None
    ) -> Dict[str, Any]:
        """Adds a click step to a StepsGraph with natural language support.
       
        Args:
            prompt: Natural language description of the click step to add (optional)
            target_graph_path: Path to the StepsGraph asset
            target_name: Target object path in format "prefabName/childName"
            step_id: Optional step ID
           
        Returns:
            Dictionary with results ('success', 'message', 'data')
        """
        return handle_add_click_step({
            "prompt": prompt,
            "target_graph_path": target_graph_path,
            "target_name": target_name,
            "step_id": step_id
        })

def handle_add_click_step(params: Dict[str, Any]) -> Dict[str, Any]:
    """
    Handle add click step tool requests.
    
    Args:
        params: Dictionary containing:
            - prompt: Natural language description (optional)
            - target_graph_path: Path to the StepsGraph asset
            - target_name: Target object path in format "prefabName/childName"
            - step_id: Optional step ID
    
    Returns:
        Dictionary containing the response from Unity
    """
    try:
        prompt = params.get('prompt')
        target_graph_path = params.get('target_graph_path')
        target_name = params.get('target_name')
        step_id = params.get('step_id')
        
        # Parse natural language prompt if provided
        if prompt and not (target_graph_path and target_name):
            # Extract graph path and target from prompt
            graph_match = re.search(r'in\s+([\w.]+\.asset)', prompt)
            target_match = re.search(r'in\s+the\s+(\w+)\s+in\s+the\s+(\w+)\s+button', prompt, re.IGNORECASE)
           
            if graph_match:
                target_graph_path = graph_match.group(1)
            if target_match:
                target_name = f"{target_match.group(1)}/{target_match.group(2)}"
               
            if not target_graph_path or not target_name:
                return {
                    "success": False,
                    "message": "Could not extract required information from prompt. Please provide explicit target_graph_path and target_name."
                }
        
        # Validate required parameters
        if not target_graph_path:
            return {"success": False, "message": "target_graph_path parameter is required"}
        
        if not target_name:
            return {"success": False, "message": "target_name parameter is required"}
            
        # Prepare parameters for Unity
        unity_params = {
            "targetGraphPath": target_graph_path,
            "targetName": target_name
        }
       
        if step_id:
            unity_params["stepId"] = step_id
            
        # Send command to Unity
        response = get_unity_connection().send_command("add_click_step", unity_params)
       
        # Process response from Unity
        if response.get("success"):
            return {
                "success": True,
                "message": "Click step added successfully",
                "data": response.get("data")
            }
        else:
            return {
                "success": False,
                "message": response.get("error", "Failed to add click step")
            }
            
    except Exception as e:
        return {
            "success": False,
            "message": f"Error adding click step: {str(e)}"
        }